src/CoreWCF.WebHttp/src/CoreWCF/OpenApi/Middleware/OpenApiMiddleware.cs

```diff
@@ -0,0 +1,106 @@
+        private readonly Dictionary<string, byte[]> _swaggerUi;
+
+        public OpenApiMiddleware(
+            RequestDelegate next,
+            OpenApiDocumentProvider openApiDocumentProvider)
+        {
+            _next = next;
+            _openApiDocumentProvider = openApiDocumentProvider;
+            _swaggerUi = new Dictionary<string, byte[]>
+            {
+                { "/swagger/favicon-16x16.png", ReadResource("favicon-16x16.png") },
+                { "/swagger/favicon-32x32.png", ReadResource("favicon-32x32.png") },
+                { "/swagger/index.html", ReadResource("index.html") },
+                { "/swagger/swagger-ui-bundle.js", ReadResource("swagger-ui-bundle.js") },
+                { "/swagger/swagger-ui-standalone-preset.js", ReadResource("swagger-ui-standalone-preset.js") },
+                { "/swagger/swagger-ui.css", ReadResource("swagger-ui.css") },
+            };
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            string requestPath = context.Request.Path.ToString();
+            if (!requestPath.StartsWith("/"))
+            {
+                requestPath = "/" + requestPath;
+            }
+
+            // Needs to happen after contracts have been added to document provider.
+            if (OpenApiDocumentationEnabled() && requestPath == "/swagger/swagger.json")
+            {
+                await WriteResponse(context, requestPath, Encoding.UTF8.GetBytes(_openApiDocumentProvider.Document));
+            }
+            else if (OpenApiDocumentationEnabled() && _swaggerUi.ContainsKey(requestPath))
+            {
+                await WriteResponse(context, requestPath, _swaggerUi[requestPath]);
+            }
+            else
+            {
+                await _next.Invoke(context).ConfigureAwait(false);
+            }
+        }
+
+        private bool OpenApiDocumentationEnabled() => _openApiDocumentProvider.Contracts.Count > 0;
+
+        private static string GetContentType(string path)
+        {
+            switch (Path.GetExtension(path))
+            {
+                case ".png":
+                    return "image/png";
+                case ".html":
+                    return "text/html";
+                case ".js":
+                    return "text/javascript";
+                case ".json":
+                    return "";
+                case ".css":
+                    return "text/css";
+                default:
+                    return null;
+            }
+        }
+
+        private async Task WriteResponse(HttpContext context, string requestPath, byte[] response)
+        {
+            context.Response.StatusCode = 200;
+            context.Response.ContentType = GetContentType(requestPath);
+            await context.Response.Body.WriteAsync(response, 0, response.Length);
+        }
```

Comment:
You know the length of the response. I believe things are better optimized if you set the ContentLength before you write the response body.